androidApp/src/main/java/com/example/tasky/android/alarm/data/AlarmScheduler.kt

```diff
@@ -4,28 +4,46 @@ import android.app.AlarmManager
 import android.app.PendingIntent
 import android.content.Context
 import android.content.Intent
+import com.example.tasky.alarm.domain.IAlarmRepository
+import com.example.tasky.alarm.domain.model.AgendaAlarm
 import com.example.tasky.android.alarm.domain.IAlarmScheduler
 import com.example.tasky.android.alarm.domain.mapper.toNotificationDataParcelable
 import com.example.tasky.android.alarm.domain.model.NotificationData
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Clock
 import kotlinx.datetime.TimeZone
 import kotlinx.datetime.toInstant
 import kotlin.time.DurationUnit
 
 class AlarmScheduler(
     private val context: Context,
+    private val alarmRepository: IAlarmRepository,
 ) : IAlarmScheduler {
     private val alarmManager = context.getSystemService(AlarmManager::class.java)
 
     override fun schedule(data: NotificationData) {
+        val triggerAtMillis =
+            data.startTime.toInstant(TimeZone.currentSystemDefault()).toEpochMilliseconds() -
+                data.remindAtType.duration.toLong(DurationUnit.MILLISECONDS)
+
+        if (triggerAtMillis <= Clock.System.now().toEpochMilliseconds()) {
+            return
+        }
+
+        CoroutineScope(Dispatchers.IO).launch {
+            alarmRepository.upsertAgendaAlarm(AgendaAlarm(data.agendaId, data.requestCode))
+        }
```

Comment:
You have no control over your scope here. Instead, change your function to be suspending to call other suspending functions.